### PR TITLE
Remove kvm dependency on memory_model

### PR DESCRIPTION
**Pull request by andreeaflorescu**
Issue #, if available: #932 

Description of changes:
Remove dependency on memory-model crate.
Refactored some functions to receive the kvm bindings structure instead of individual parameters.
Refactored kvm get dirty log to not assume that the memory size is a multiple of 256KB.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: kvm/src/lib.rs [hunk 1]
```diff
@@ -984,6 +1013,34 @@ mod tests {
         }
     }
 
+    // Helper function for mmap an anonymous memory of `size`.
+    // Panics if the mmap fails.
+    fn mmap_anonymous(size: usize) -> *mut u8 {
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply kvm/src/lib.rs]: A mmap helper in from_fd would be use in just one place. I think that function is small enough to not be further split.

File: kvm/src/lib.rs [hunk 2]
```diff
@@ -1004,6 +1013,34 @@ mod tests {
...
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            panic!("mmap failed.");
+        }
+
+        return addr as *mut u8;
+    }
+
+    impl KvmRunWrapper {
```
Comment (dianpopa) [inline kvm/src/lib.rs]: Given that this is used only in the test that checks the behavior in case of a faulty vcpu fd, it is not necessary to implement another KvmRunWrapper. See my suggestion from the test where this is used.

File: kvm/src/lib.rs [hunk 3]
```diff
@@ -1474,7 +1518,7 @@ mod tests {
         assert_eq!(get_raw_errno(faulty_vm_fd.create_vcpu(0)), badf_errno);
         let faulty_vcpu_fd = VcpuFd {
             vcpu: unsafe { File::from_raw_fd(-1) },
-            run_mmap: MemoryMapping::new(10).unwrap(),
+            kvm_run_ptr: KvmRunWrapper::new(10),
```
Comment (dianpopa) [inline kvm/src/lib.rs]:
```suggestion
            kvm_run_ptr: KvmRunWrapper {
                kvm_run_ptr: 0x0 as *mut u8,
            },
```
There is no need to implement another KvmRunWrapper given that this is concerned only with a faulty vcpu fd.

File: kvm/src/lib.rs [hunk 4]
```diff
@@ -430,16 +463,16 @@ impl VmFd {
...
-        );
-        // Create a vec with at least one bit per page of guest memory in this slot.
-        let bitmap_size = memory_size / memsize_multiplier;
+        // Compute the length of the bitmap needed for all dirty pages in one memory slot.
+        // One memory page is 4KiB (4096 bits) and KVM_GET_DIRTY_LOG returns one dirty bit for
+        // each page.
+        let page_size = 4 << 10;
+
+        let div_round_up = |dividend, divisor| (dividend + divisor - 1) / divisor;
+        // For ease of access we are saving the bitmap in a u64 vector. We are using ceil to
```
Comment (dianpopa) [inline kvm/src/lib.rs]: From the comment it is not clear that we are using ceil to calculate the capacity of the `u64` vector. I would rephrase it.
```suggestion
        // For ease of access we are saving the bitmap in an `u64` vector. We are calculating its capacity by dividing the total memory to the size of a page. Even if `mem_size` is not a multiple of `page_size * 64` we make sure all dirty pages are counted by using a ceiling approximation.
```

File: kvm/src/lib.rs (hunk 2 above)
Comment (andreeaflorescu) [reply kvm/src/lib.rs]: This is not used only in tests. The need for a KvmRunWrapper is detailed in the comment above the structure definition. Specifically, we need a wrapper and we cannot use a raw pointer (defined as either *mut u8 or *mut kvm_run) because when we spawn the vcpu threads the VcpuFd structure is moved. To safely move a structure between threads all the fields of the structure need to implement Send and Sync.

Comment (dianpopa) [reply kvm/src/lib.rs]: > This is not used only in tests. The need for a KvmRunWrapper is detailed in the comment above the structure definition. Specifically, we need a wrapper and we cannot use a raw pointer (defined as either *mut u8 or *mut kvm_run) because when we spawn the vcpu threads the VcpuFd structure is moved. To safely move a structure between threads all the fields of the structure need to implement Send and Sync.

 I see *two* `impl KvmRunWrapper` inside the kvm/src/lib.rs. One that is used for the reasons that you explained and another one used only in tests. Can you please double check?

Comment (andreeaflorescu) [reply kvm/src/lib.rs]: Ah ok, I get it now. So, the `KvmRunWrapper` is not another implementation for `KvmRunWrapper`, it is just an extension of the existing `KvmRunWrapper`. Basically, on top of the existing methods, we also add a new method.

The reason I decided against using a hardcoded address for kvm_run like you did in the example is because the test is checking the errors in case you have a bad file descriptor for either Kvm, VmFd or VcpuFd. If we also pass an invalid pointer (that is not mapped) for kvm_run then we are altering 2 fields and we cannot be sure which one of them generated the errors.
